File: apps/server/integrations/vtube_studio.py
"""
VTube Studio WebSocket API Client

Connects to VTube Studio via WebSocket and controls avatar parameters and expressions.
Authentication tokens are cached to file for persistent sessions.
"""

import asyncio
import json
import logging
from pathlib import Path
from typing import Optional, Callable, Awaitable
from datetime import datetime

import websockets
from websockets.client import WebSocketClientProtocol

logger = logging.getLogger(__name__)

# VTS API constants
PLUGIN_NAME = "AITuberFlow"
PLUGIN_DEVELOPER = "AITuberFlow"
API_NAME = "VTubeStudioPublicAPI"
API_VERSION = "1.0"

# Token storage path
TOKEN_FILE = Path(__file__).parent.parent / "data" / "vts_token.json"


class VTubeStudioClient:
    """
    VTube Studio WebSocket API client.

    Handles connection, authentication, and parameter/hotkey control.
    """

    def __init__(self):
        self.ws: Optional[WebSocketClientProtocol] = None
        self.port: int = 8001
        self.connected: bool = False
        self.authenticated: bool = False
        self.request_id: int = 0
        self.pending_requests: dict[str, asyncio.Future] = {}
        self.expression_map: dict[str, str] = {}
        self.mouth_param: str = "MouthOpen"
        self._receive_task: Optional[asyncio.Task] = None
        self._reconnect_task: Optional[asyncio.Task] = None
        self._should_reconnect: bool = True

    @property
    def is_connected(self) -> bool:
        """Check if connected and authenticated."""
        return self.connected and self.authenticated

    def configure(
        self,
        port: int = 8001,
        mouth_param: str = "MouthOpen",
        expression_map: Optional[dict[str, str]] = None
    ):
        """Configure VTS client settings."""
        self.port = port
        self.mouth_param = mouth_param
        self.expression_map = expression_map or {
            "happy": "Happy",
            "sad": "Sad",
            "angry": "Angry",
            "surprised": "Surprised",
            "relaxed": "Relaxed",
            "neutral": "Neutral",
        }
        logger.info(f"VTS configured: port={port}, mouth_param={mouth_param}")

    async def connect(self) -> bool:
        """
        Connect to VTube Studio and authenticate.

        Returns True if connection and authentication successful.
        """
        if self.connected:
            return self.authenticated

        self._should_reconnect = True

        try:
            uri = f"ws://localhost:{self.port}"
            logger.info(f"Connecting to VTube Studio at {uri}")

            self.ws = await websockets.connect(uri)
            self.connected = True
            logger.info("Connected to VTube Studio")

            # Start receive loop
            self._receive_task = asyncio.create_task(self._receive_loop())

            # Authenticate
            success = await self._authenticate()
            if success:
                logger.info("VTube Studio authentication successful")
            else:
                logger.warning("VTube Studio authentication failed")

            return success

        except Exception as e:
            logger.error(f"Failed to connect to VTube Studio: {e}")
            self.connected = False
            self.authenticated = False
            return False

    async def disconnect(self):
        """Disconnect from VTube Studio."""
        self._should_reconnect = False

        if self._reconnect_task:
            self._reconnect_task.cancel()
            self._reconnect_task = None

        if self._receive_task:
            self._receive_task.cancel()
            self._receive_task = None

        if self.ws:
            await self.ws.close()
            self.ws = None

        self.connected = False
        self.authenticated = False
        logger.info("Disconnected from VTube Studio")

    async def set_parameter(self, param_id: str, value: float) -> bool:
        """
        Set a VTS parameter value (e.g., for lip sync).

        Args:
            param_id: Parameter ID (e.g., "MouthOpen")
            value: Value between 0.0 and 1.0

        Returns True if successful.
        """
        if not self.is_connected:
            return False

        value = max(0.0, min(1.0, value))

        try:
            await self._send_request(
                "InjectParameterDataRequest",
                {
                    "faceFound": True,
                    "mode": "set",
                    "parameterValues": [
                        {"id": param_id, "value": value}
                    ]
                }
            )
            return True
        except Exception as e:
            # Silently fail for high-frequency parameter updates
            return False

    async def set_mouth_open(self, value: float) -> bool:
        """Set mouth open parameter for lip sync."""
        return await self.set_parameter(self.mouth_param, value)

    async def trigger_hotkey(self, hotkey_id: str) -> bool:
        """
        Trigger a VTS hotkey (e.g., for expression change).

        Args:
            hotkey_id: Hotkey ID or name

        Returns True if successful.
        """
        if not self.is_connected:
            return False

        try:
            response = await self._send_request(
                "HotkeyTriggerRequest",
                {"hotkeyID": hotkey_id}
            )
            return True
        except Exception as e:
            logger.error(f"Failed to trigger hotkey {hotkey_id}: {e}")
            return False

    async def trigger_expression(self, expression: str) -> bool:
        """
        Trigger expression change using mapped hotkey.

        Args:
            expression: Expression name (e.g., "happy")

        Returns True if successful.
        """
        hotkey_id = self.expression_map.get(expression.lower())
        if not hotkey_id:
            logger.warning(f"No hotkey mapped for expression: {expression}")
            return False

        return await self.trigger_hotkey(hotkey_id)

    async def _authenticate(self) -> bool:
        """Authenticate with VTube Studio."""
        # Try cached token first
        token = self._load_token()

        if token:
            success = await self._auth_with_token(token)
            if success:
                return True
            # Token invalid, clear it
            self._clear_token()

        # Request new token
        token = await self._request_token()
        if not token:
            return False

        # Authenticate with new token
        success = await self._auth_with_token(token)
        if success:
            self._save_token(token)

        return success

    async def _request_token(self) -> Optional[str]:
        """Request authentication token from VTS (triggers user popup)."""
        try:
            logger.info("Requesting VTS authentication token (check VTube Studio for popup)")
            response = await self._send_request(
                "AuthenticationTokenRequest",
                {
                    "pluginName": PLUGIN_NAME,
                    "pluginDeveloper": PLUGIN_DEVELOPER,
                },
                timeout=60.0  # User needs time to click
            )

            if response and "data" in response:
                return response["data"].get("authenticationToken")
            return None

        except Exception as e:
            logger.error(f"Token request failed: {e}")
            return None

    async def _auth_with_token(self, token: str) -> bool:
        """Authenticate with existing token."""
        try:
            response = await self._send_request(
                "AuthenticationRequest",
                {
                    "pluginName": PLUGIN_NAME,
                    "pluginDeveloper": PLUGIN_DEVELOPER,
                    "authenticationToken": token,
                }
            )

            if response and "data" in response:
                self.authenticated = response["data"].get("authenticated", False)
                return self.authenticated
            return False

        except Exception as e:
            logger.error(f"Authentication failed: {e}")
            return False

    def _load_token(self) -> Optional[str]:
        """Load cached token from file."""
        try:
            if TOKEN_FILE.exists():
                data = json.loads(TOKEN_FILE.read_text())
                # Check if token is for same port
                if data.get("port") == self.port:
                    return data.get("token")
        except Exception as e:
            logger.debug(f"Failed to load token: {e}")
        return None

    def _save_token(self, token: str):
        """Save token to file."""
        try:
            TOKEN_FILE.parent.mkdir(parents=True, exist_ok=True)
            TOKEN_FILE.write_text(json.dumps({
                "token": token,
                "port": self.port,
                "updated_at": datetime.utcnow().isoformat()
            }, indent=2))
            logger.info("VTS token saved")
        except Exception as e:
            logger.error(f"Failed to save token: {e}")

    def _clear_token(self):
        """Clear cached token."""
        try:
            if TOKEN_FILE.exists():
                TOKEN_FILE.unlink()
        except Exception:
            pass

    async def _send_request(
        self,
        message_type: str,
        data: Optional[dict] = None,
        timeout: float = 10.0
    ) -> Optional[dict]:
        """Send request and wait for response."""
        if not self.ws:
            raise RuntimeError("Not connected")

        self.request_id += 1
        request_id = f"req_{self.request_id}_{int(datetime.utcnow().timestamp() * 1000)}"

        message = {
            "apiName": API_NAME,
            "apiVersion": API_VERSION,
            "requestID": request_id,
            "messageType": message_type,
        }
        if data:
            message["data"] = data

        # Create future for response
        future: asyncio.Future = asyncio.Future()
        self.pending_requests[request_id] = future

        try:
            await self.ws.send(json.dumps(message))
            return await asyncio.wait_for(future, timeout=timeout)
        except asyncio.TimeoutError:
            logger.warning(f"Request {message_type} timed out")
            return None
        finally:
            self.pending_requests.pop(request_id, None)

    async def _receive_loop(self):
        """Receive and dispatch messages from VTS."""
        try:
            async for message in self.ws:
                try:
                    data = json.loads(message)
                    request_id = data.get("requestID")

                    if request_id and request_id in self.pending_requests:
                        self.pending_requests[request_id].set_result(data)

                except json.JSONDecodeError:
                    logger.warning("Failed to parse VTS message")

        except websockets.exceptions.ConnectionClosed:
            logger.info("VTS connection closed")
            self.connected = False
            self.authenticated = False

            # Attempt reconnect
            if self._should_reconnect:
                self._reconnect_task = asyncio.create_task(self._reconnect())

        except Exception as e:
            logger.error(f"VTS receive error: {e}")
            self.connected = False
            self.authenticated = False

    async def _reconnect(self):
        """Attempt to reconnect after disconnection."""
        await asyncio.sleep(3.0)

        if self._should_reconnect and not self.connected:
            logger.info("Attempting to reconnect to VTS...")
            await self.connect()


# Global client instance
vts_client = VTubeStudioClient()

File: apps/server-ts/src/integrations/vtube-studio.ts
// Bun provides global WebSocket (browser-standard API) - no external package needed
import { readFileSync, writeFileSync, mkdirSync, unlinkSync, existsSync } from "node:fs";
import { dirname, resolve } from "node:path";

const PLUGIN_NAME = "AITuberFlow";
const PLUGIN_DEVELOPER = "AITuberFlow";
const API_NAME = "VTubeStudioPublicAPI";
const API_VERSION = "1.0";

const TOKEN_FILE = resolve(import.meta.dir ?? ".", "../../data/vts_token.json");

interface PendingRequest {
  resolve: (value: unknown) => void;
  reject: (reason: unknown) => void;
  timer: ReturnType<typeof setTimeout>;
}

class VTubeStudioClient {
  private ws: WebSocket | null = null;
  private port = 8001;
  connected = false;
  authenticated = false;
  private requestId = 0;
  private pendingRequests = new Map<string, PendingRequest>();
  private expressionMap: Record<string, string> = {};
  private mouthParam = "MouthOpen";
  private shouldReconnect = true;
  private abortController: AbortController | null = null;

  get isConnected(): boolean {
    return this.connected && this.authenticated;
  }

  configure(
    port = 8001,
    mouthParam = "MouthOpen",
    expressionMap?: Record<string, string>
  ): void {
    this.port = port;
    this.mouthParam = mouthParam;
    this.expressionMap = expressionMap ?? {
      happy: "Happy",
      sad: "Sad",
      angry: "Angry",
      surprised: "Surprised",
      relaxed: "Relaxed",
      neutral: "Neutral",
    };
    console.log(`[VTS] Configured: port=${port}, mouthParam=${mouthParam}`);
  }

  async connect(): Promise<boolean> {
    if (this.connected) return this.authenticated;

    this.shouldReconnect = true;

    try {
      const uri = `ws://localhost:${this.port}`;
      console.log(`[VTS] Connecting to VTube Studio at ${uri}`);

      return await new Promise<boolean>((resolve) => {
        const ws = new WebSocket(uri);
        this.abortController = new AbortController();
        const { signal } = this.abortController;

        ws.addEventListener("open", async () => {
          this.ws = ws;
          this.connected = true;
          console.log("[VTS] Connected to VTube Studio");
          this.setupReceiveHandler();

          const success = await this.authenticate();
          if (success) {
            console.log("[VTS] Authentication successful");
          } else {
            console.warn("[VTS] Authentication failed");
          }
          resolve(success);
        }, { signal });

        ws.addEventListener("error", () => {
          console.error("[VTS] Connection error");
          this.connected = false;
          this.authenticated = false;
          resolve(false);
        }, { signal });
      });
    } catch (e) {
      console.error(`[VTS] Failed to connect: ${e}`);
      this.connected = false;
      this.authenticated = false;
      return false;
    }
  }

  async disconnect(): Promise<void> {
    this.shouldReconnect = false;

    for (const [id, req] of this.pendingRequests) {
      clearTimeout(req.timer);
      req.reject(new Error("Disconnecting"));
      this.pendingRequests.delete(id);
    }

    if (this.ws) {
      this.abortController?.abort();
      this.abortController = null;
      this.ws.close();
      this.ws = null;
    }

    this.connected = false;
    this.authenticated = false;
    console.log("[VTS] Disconnected from VTube Studio");
  }

  async setParameter(paramId: string, value: number): Promise<boolean> {
    if (!this.isConnected) return false;

    value = Math.max(0, Math.min(1, value));

    try {
      await this.sendRequest("InjectParameterDataRequest", {
        faceFound: true,
        mode: "set",
        parameterValues: [{ id: paramId, value }],
      });
      return true;
    } catch {
      return false;
    }
  }

  async setMouthOpen(value: number): Promise<boolean> {
    return this.setParameter(this.mouthParam, value);
  }

  async triggerHotkey(hotkeyId: string): Promise<boolean> {
    if (!this.isConnected) return false;

    try {
      await this.sendRequest("HotkeyTriggerRequest", { hotkeyID: hotkeyId });
      return true;
    } catch (e) {
      console.error(`[VTS] Failed to trigger hotkey ${hotkeyId}: ${e}`);
      return false;
    }
  }

  async triggerExpression(expression: string): Promise<boolean> {
    const hotkeyId = this.expressionMap[expression.toLowerCase()];
    if (!hotkeyId) {
      console.warn(`[VTS] No hotkey mapped for expression: ${expression}`);
      return false;
    }
    return this.triggerHotkey(hotkeyId);
  }

  // ─── Authentication ─────────────────────────────────────────────

  private async authenticate(): Promise<boolean> {
    const token = this.loadToken();

    if (token) {
      const success = await this.authWithToken(token);
      if (success) return true;
      this.clearToken();
    }

    const newToken = await this.requestToken();
    if (!newToken) return false;

    const success = await this.authWithToken(newToken);
    if (success) this.saveToken(newToken);
    return success;
  }

  private async requestToken(): Promise<string | null> {
    try {
      console.log("[VTS] Requesting auth token (check VTube Studio for popup)");
      const response = (await this.sendRequest(
        "AuthenticationTokenRequest",
        {
          pluginName: PLUGIN_NAME,
          pluginDeveloper: PLUGIN_DEVELOPER,
        },
        60_000
      )) as Record<string, any> | null;

      return response?.data?.authenticationToken ?? null;
    } catch (e) {
      console.error(`[VTS] Token request failed: ${e}`);
      return null;
    }
  }

  private async authWithToken(token: string): Promise<boolean> {
    try {
      const response = (await this.sendRequest("AuthenticationRequest", {
        pluginName: PLUGIN_NAME,
        pluginDeveloper: PLUGIN_DEVELOPER,
        authenticationToken: token,
      })) as Record<string, any> | null;

      if (response?.data) {
        this.authenticated = response.data.authenticated === true;
        return this.authenticated;
      }
      return false;
    } catch (e) {
      console.error(`[VTS] Authentication failed: ${e}`);
      return false;
    }
  }

  // ─── Token persistence ──────────────────────────────────────────

  private loadToken(): string | null {
    try {
      if (existsSync(TOKEN_FILE)) {
        const data = JSON.parse(readFileSync(TOKEN_FILE, "utf-8"));
        if (data.port === this.port) return data.token ?? null;
      }
    } catch {
      // ignore
    }
    return null;
  }

  private saveToken(token: string): void {
    try {
      mkdirSync(dirname(TOKEN_FILE), { recursive: true });
      writeFileSync(
        TOKEN_FILE,
        JSON.stringify(
          { token, port: this.port, updatedAt: new Date().toISOString() },
          null,
          2
        )
      );
      console.log("[VTS] Token saved");
    } catch (e) {
      console.error(`[VTS] Failed to save token: ${e}`);
    }
  }

  private clearToken(): void {
    try {
      if (existsSync(TOKEN_FILE)) unlinkSync(TOKEN_FILE);
    } catch {
      // ignore
    }
  }

  // ─── WebSocket transport ────────────────────────────────────────

  private sendRequest(
    messageType: string,
    data?: Record<string, unknown>,
    timeout = 10_000
  ): Promise<unknown> {
    return new Promise((resolve, reject) => {
      if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
        return reject(new Error("Not connected"));
      }

      this.requestId += 1;
      const requestID = `req_${this.requestId}_${Date.now()}`;

      const message: Record<string, unknown> = {
        apiName: API_NAME,
        apiVersion: API_VERSION,
        requestID,
        messageType,
      };
      if (data) message.data = data;

      const timer = setTimeout(() => {
        this.pendingRequests.delete(requestID);
        console.warn(`[VTS] Request ${messageType} timed out`);
        reject(new Error(`Request ${messageType} timed out`));
      }, timeout);

      this.pendingRequests.set(requestID, { resolve, reject, timer });
      this.ws.send(JSON.stringify(message));
    });
  }

  private setupReceiveHandler(): void {
    if (!this.ws || !this.abortController) return;

    const { signal } = this.abortController;

    this.ws.addEventListener("message", (event: MessageEvent) => {
      try {
        const data = JSON.parse(
          typeof event.data === "string" ? event.data : String(event.data)
        ) as Record<string, unknown>;
        const requestId = data.requestID as string | undefined;

        if (requestId && this.pendingRequests.has(requestId)) {
          const pending = this.pendingRequests.get(requestId)!;
          clearTimeout(pending.timer);
          this.pendingRequests.delete(requestId);
          pending.resolve(data);
        }
      } catch {
        console.warn("[VTS] Failed to parse message");
      }
    }, { signal });

    this.ws.addEventListener("close", () => {
      console.log("[VTS] Connection closed");
      this.connected = false;
      this.authenticated = false;

      if (this.shouldReconnect) {
        setTimeout(() => this.reconnect(), 3000);
      }
    }, { signal });

    this.ws.addEventListener("error", () => {
      console.error("[VTS] WebSocket error");
      this.connected = false;
      this.authenticated = false;
    }, { signal });
  }

  private async reconnect(): Promise<void> {
    if (this.shouldReconnect && !this.connected) {
      console.log("[VTS] Attempting to reconnect...");
      await this.connect();
    }
  }
}

export const vtsClient = new VTubeStudioClient();
